### DISCUSSION TITLE
Set sensor values via API

### DISCUSSION THREAD

**Discussion by CommandoThanatos** [goecharger/go-eController-API#2] [Ideas]
Would be great if it there would be a possibility to set the energy value of a sensor via the api.
Why? I own a smart home system which measures the production of a mini photovoltaic system via a "Fritz Schaltsteckdose" and there are also other devices which i would like to be recognized by the  go-e controller anzeigen and to add them to kategories (like solar with valiues of the big pv and the mini pv) to show the real production for all systems and to add the mini pv to the overall consumption of the house.

**Comment by SIMULATAN** [comment]
Yes, absolutely.

As someone with the go-e controller and a DC battery, this is an absolute necessity.
(only AC batteries can be read, there is no possibility of reading DC ones - see [the go-e magazine](https://go-e.com/en/magazine/go-e-controller-explained))
Without the ability to ingest power consumption and battery state from arbitrary sources, the controller unfortunately is useless in a lot of cases when it comes to excess charging.

A note regarding the implementation: it'd be great to implement the setters in the "PATCH"-style, as in, you can choose to update only a subset of the values. This would allow multiple sources writing data concurrently leading to further flexibility.
I'd personally choose MQTT as the leading protocol for these high frequency values as it's made for exactly this use case. Something similar to the go-e charger would be perfectly applicable.

Thanks for the otherwise great products!

**Reply by tom-mi** [reply]
So are there any updates on this @0xFEEDC0DE64  @erxbout ?
I currently have one charger in place which I feed via Homeassistant with realtime battery & power readings (which I get from my inverter). I bought a controller for load balancing, as I will install a second charger soon. I expected to be able to just point my automation to publishing to the controller, but as this doesn't seem to work, I'm quite disappointed now that
1) I actually would have to connect the sensors of the controller (even though my inverter already provides all the values collected via it's own sensors)
2) I would miss the battery data

So using the controller in its current state would be a downgrade regarding data quality, but seems necessary to get some sort of proper load balancing working (?)

**Reply by tom-mi** [reply]
Ok ich nehme alles zurück und behaupte das Gegenteil - I _think_ I got it working (not tested with real car yet) using the instructions in this comment
https://github.com/goecharger/go-eController-API/issues/14#issuecomment-2595713130
and publishing that 16-element array via MQTT to `.../ecp/set`. 
There is this caveat though
https://github.com/goecharger/go-eController-API/issues/14#issuecomment-2599267728
> Please note that this might break if the position of the battery power changes because of a firmware update or other changing settings so fetching the index before via 'ccn' would be better practice as stated in my clarification

That would be much more complex than just using the builtin mqtt publish action in Homeassistant. I'd love to see
* this to be documented somewhere :) 
* an API similar to the one for the charger (sending a dict / object instead of a "magic" array)

**Reply by tom-mi** [reply]
So as an intermediate result about what is working / not working right now (tested with controller + 1 charger):
:heavy_check_mark: setting power for home, grid, pv, battery via `.../ecp/set` (however, I had to deassociate the internal sensors for the power values not to be falsified)
:heavy_check_mark: => PV surplus charging with 1 charger (power values sent via Inverter -> Home Assistant -> Controller -> Charger)
:x: setting current values for the grid via API (I could not find any API key/topic that would allow that)
:x: => dynamic loadbalancing setting seems to have no effect, as the charger does not reduce charging power to compensate power drawn from the grid for other consumers in the house (my assumption is that this is due to the missing current values)
:grey_question: loadbalancing between multiple chargers not tested yet

**Comment by SvenAbels** [comment]
I totally agree with this suggestion. - Especially for DC battery status delivery

**Comment by norcon** [comment]
It would be great to set pGrid, pSolar, pAkku, … via MQTT or other APIs like in Charger API

**Comment by 0xFEEDC0DE64** [comment]
The next firmware release will have the energy counters writable on the go-e controller

**Reply by norcon** [reply]
Not available in 1.0.8 Beta right? 
Tested with Mqtt:
unknown api key: ids

**Comment by mi-hol** [comment]
> The next firmware release

With missing documentation for MQTT as noted in https://github.com/goecharger/go-eController-API/issues/3 how can the advertised MQTT feature be used/tested?

**Reply by erxbout** [reply]
Protocoll for using mqtt is the same as https://github.com/goecharger/go-eCharger-API-v2/blob/main/mqtt-en.md
The api keys for the controller are here https://github.com/goecharger/go-eController-API/blob/main/apikeys-en.md

I´ll update the documentation as soon as possible and also add mqtt docs to controller repo

**Reply by mi-hol** [reply]
> I´ll update the documentation as soon as possible

Just to be sure, by above do you mean to resolve "TODO for go-e: Document the topics for reading/writing."?

**Comment by JaroRobo3000** [comment]
It would also be very good if, instead of the sensor inputs for measuring the individual phases, it were possible to make these values writable via the API, for example, to make them "isv" writable. I have a Victron system with an energy meter, but it is connected to the GX device via USB. So I would have to transfer the values via Node Red.

**Reply by norcon** [reply]
Same here!
